crates/contract/src/event.rs

```diff
@@ -131,10 +140,67 @@ fn decode_log<E: SolEvent>(log: &Log) -> alloy_sol_types::Result<E> {
     E::decode_raw_log(log.topics.iter().copied(), &log.data, false)
 }
 
+#[cfg(feature = "pubsub")]
+pub(crate) mod subscription {
+    use super::*;
+    use alloy_pubsub::Subscription;
+
+    /// An event subscription.
+    ///
+    /// Underlying subscription is available through the [`sub`](Self::sub) field.
+    pub struct EventSubscription<E> {
+        /// The inner poller.
+        pub sub: Subscription<Log>,
+        _phantom: PhantomData<E>,
+    }
+
+    impl<E> std::ops::Deref for EventSubscription<E> {
+        type Target = Subscription<Log>;
+
+        #[inline]
+        fn deref(&self) -> &Self::Target {
+            &self.sub
+        }
+    }
+
+    impl<E> std::ops::DerefMut for EventSubscription<E> {
+        #[inline]
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            &mut self.sub
+        }
+    }
+
+    impl<E> fmt::Debug for EventSubscription<E> {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            f.debug_struct("EventSubscription")
+                .field("sub", &self.sub)
+                .field("event_type", &format_args!("{}", std::any::type_name::<E>()))
+                .finish()
+        }
+    }
+
+    impl<E: SolEvent> EventSubscription<E> {
+        /// Creates a new event subscription with the provided subscription.
+        #[allow(clippy::missing_const_for_fn)]
+        #[inline]
+        pub fn new(sub: Subscription<Log>) -> Self {
```

Comment:
`From` impl preferred here

Reply:
added for both `EventPoller` and `EventSubscription` in d37f0503918e7a7d9e953e664c3b4293f25a1bad